File: backends/spring-boot/src/main/java/com/hellotime/controller/AdminController.java
package com.hellotime.controller;

import com.hellotime.dto.*;
import com.hellotime.exception.UnauthorizedException;
import com.hellotime.service.AdminService;
import com.hellotime.service.CapsuleService;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;

/**
 * 管理员控制器
 * 处理管理员相关的 HTTP 请求
 * 基础路径：/api/v1/admin
 * 所有接口需要 Bearer Token 认证（除登录外）
 */
@RestController
@RequestMapping("/api/v1/admin")
public class AdminController {

    private final AdminService adminService;
    private final CapsuleService capsuleService;

    /**
     * 构造函数注入依赖
     */
    public AdminController(AdminService adminService, CapsuleService capsuleService) {
        this.adminService = adminService;
        this.capsuleService = capsuleService;
    }

    /**
     * 管理员登录
     * POST /api/v1/admin/login
     *
     * @param request 登录请求（包含密码）
     * @return JWT Token
     * @throws UnauthorizedException 密码错误时抛出
     */
    @PostMapping("/login")
    public ApiResponse<AdminTokenResponse> login(@Valid @RequestBody AdminLoginRequest request) {
        String token = adminService.login(request.getPassword());
        if (token == null) {
            throw new UnauthorizedException("密码错误");
        }
        return ApiResponse.ok(new AdminTokenResponse(token), "登录成功");
    }

    /**
     * 分页查询所有胶囊
     * GET /api/v1/admin/capsules?page=0&size=20
     * 需要认证
     *
     * @param page 页码（默认 0）
     * @param size 每页大小（默认 20）
     * @return 胶囊列表分页数据
     */
    @GetMapping("/capsules")
    public ApiResponse<PageResponse<CapsuleResponse>> list(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        return ApiResponse.ok(capsuleService.listCapsules(page, size));
    }

    /**
     * 删除指定胶囊
     * DELETE /api/v1/admin/capsules/{code}
     * 需要认证
     *
     * @param code 8 位胶囊码
     * @return 删除结果
     */
    @DeleteMapping("/capsules/{code}")
    public ApiResponse<Void> delete(@PathVariable String code) {
        capsuleService.deleteCapsule(code);
        return ApiResponse.ok(null, "删除成功");
    }
}

File: backends/nest/src/admin/admin.controller.ts
/**
 * 管理控制器 — 处理 /admin 路由（登录、列表、删除）
 *
 * @UseGuards(AdminAuthGuard) 将守卫挂载到需要认证的路由上，
 * 比 Spring Boot 的 @PreAuthorize 更显式，比 Gin 的 router.Group 更细粒度。
 * 登录路由不需要守卫，其他两个路由都需要。
 */
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  Post,
  Query,
  UseGuards,
} from '@nestjs/common';
/**
 * 管理控制器 — 处理 /admin 路由（登录、列表、删除）
 *
 * @UseGuards(AdminAuthGuard) 将守卫挂载到需要认证的路由上，
 * 比 Spring Boot 的 @PreAuthorize 更显式，比 Gin 的 router.Group 更细粒度。
 * 登录路由不需要守卫，其他两个路由都需要。
 */
import { ok } from '../common/dto/api-response.dto';
/**
 * 管理控制器 — 处理 /admin 路由（登录、列表、删除）
 *
 * @UseGuards(AdminAuthGuard) 将守卫挂载到需要认证的路由上，
 * 比 Spring Boot 的 @PreAuthorize 更显式，比 Gin 的 router.Group 更细粒度。
 * 登录路由不需要守卫，其他两个路由都需要。
 */
import { AppUnauthorizedException } from '../common/exceptions/app-unauthorized.exception';
/**
 * 管理控制器 — 处理 /admin 路由（登录、列表、删除）
 *
 * @UseGuards(AdminAuthGuard) 将守卫挂载到需要认证的路由上，
 * 比 Spring Boot 的 @PreAuthorize 更显式，比 Gin 的 router.Group 更细粒度。
 * 登录路由不需要守卫，其他两个路由都需要。
 */
import { CapsulesService } from '../capsules/capsules.service';
/**
 * 管理控制器 — 处理 /admin 路由（登录、列表、删除）
 *
 * @UseGuards(AdminAuthGuard) 将守卫挂载到需要认证的路由上，
 * 比 Spring Boot 的 @PreAuthorize 更显式，比 Gin 的 router.Group 更细粒度。
 * 登录路由不需要守卫，其他两个路由都需要。
 */
import { AdminAuthGuard } from './admin-auth.guard';
/**
 * 管理控制器 — 处理 /admin 路由（登录、列表、删除）
 *
 * @UseGuards(AdminAuthGuard) 将守卫挂载到需要认证的路由上，
 * 比 Spring Boot 的 @PreAuthorize 更显式，比 Gin 的 router.Group 更细粒度。
 * 登录路由不需要守卫，其他两个路由都需要。
 */
import { AdminService } from './admin.service';
/**
 * 管理控制器 — 处理 /admin 路由（登录、列表、删除）
 *
 * @UseGuards(AdminAuthGuard) 将守卫挂载到需要认证的路由上，
 * 比 Spring Boot 的 @PreAuthorize 更显式，比 Gin 的 router.Group 更细粒度。
 * 登录路由不需要守卫，其他两个路由都需要。
 */
import { AdminLoginDto } from './dto/admin-login.dto';
/**
 * 管理控制器 — 处理 /admin 路由（登录、列表、删除）
 *
 * @UseGuards(AdminAuthGuard) 将守卫挂载到需要认证的路由上，
 * 比 Spring Boot 的 @PreAuthorize 更显式，比 Gin 的 router.Group 更细粒度。
 * 登录路由不需要守卫，其他两个路由都需要。
 */
import { ListCapsulesQueryDto } from './dto/list-capsules-query.dto';

@Controller('admin')
export class AdminController {
  constructor(
    private readonly adminService: AdminService,
    private readonly capsulesService: CapsulesService,
  ) {}

  @Post('login')
  @HttpCode(HttpStatus.OK)
  login(@Body() request: AdminLoginDto) {
    const token = this.adminService.login(request.password);
    if (!token) {
      throw new AppUnauthorizedException('密码错误');
    }
    return ok({ token }, '登录成功');
  }

  @Get('capsules')
  @UseGuards(AdminAuthGuard)
  list(@Query() query: ListCapsulesQueryDto) {
    return ok(this.capsulesService.listCapsules(query.page, query.size));
  }

  @Delete('capsules/:code')
  @UseGuards(AdminAuthGuard)
  remove(@Param('code') code: string) {
    this.capsulesService.deleteCapsule(code);
    return ok(null, '删除成功');
  }
}
